### PR TITLE
fix: pass through redirect query parameters

### PR DESCRIPTION
**Pull request by J0**
## What kind of change does this PR introduce?

Aims to address #1150  so that we can pass redirect query parameters alongside errors. Note that if any of the existing query parameters are named `error` or `error_description` they will be overwritten. In such cases, the error added by Supabase Auth will take precedence

## What is the current behavior?

When an error occurs the redirect query params only return error and error_description without other query parameters.

## How this was tested
- Case with Error was induced by artifically returning an error in `internalExternalProviderCallback`
- Happy path was tested as per normal

### CODE REVIEW

**Comment by J0** [comment]
For Testing

File: internal/api/external.go [hunk 1]
```diff
@@ -587,7 +592,7 @@ func getErrorQueryString(err error, errorID string, log logrus.FieldLogger) *url
 		q.Set("error_description", e.Description)
 		log.WithError(e.Cause()).Info(e.Error())
 	case ErrorCause:
-		return getErrorQueryString(e.Cause(), errorID, log)
+		return getErrorQueryString(e.Cause(), errorID, log, q)
```
Comment (hf) [inline internal/api/external.go:595]: Is this recursive? Why is this necessary to be recursive?

Comment (J0) [reply internal/api/external.go:595]: yeah it came as a recursive function - not sure why it was designed as such

Comment (kangmingtay) [reply internal/api/external.go:595]: this is fine though right? it's a tail recursion so no perf implications

**Comment by kangmingtay** [comment]
> Not that if any of the existing redirect errors are named error or error_description they will be overwritten.

just to be clear, i think you mean "if any of the existing _query parameters_ are named error or error_description they will be overwritten" right?

**Comment by J0** [comment]
Amended the description as per suggestion . Thanks!

File: internal/api/external.go [hunk 2]
```diff
@@ -554,19 +559,19 @@ func (a *API) Provider(ctx context.Context, name string, scopes string) (provide
...
 	errorID := getRequestID(ctx)
 	err := handler(w, r)
 	if err != nil {
-		q := getErrorQueryString(err, errorID, log)
-		http.Redirect(w, r, a.getExternalRedirectURL(r)+"?"+q.Encode(), http.StatusFound)
+		q := getErrorQueryString(err, errorID, log, u.Query())
+		u.RawQuery = q.Encode()
+		http.Redirect(w, r, u.String(), http.StatusFound)
 	}
 }
```
Comment (kangmingtay) [inline internal/api/external.go:572]: can we add a unit test too for the `redirectErrors` function to prevent this from regressing please?

**Review by kangmingtay — APPROVED** [review]

File: internal/api/external.go (hunk 1 above)
Comment (J0) [reply internal/api/external.go:595]: Going to mark as resolve as it shouldn't cause any significant performance decrease. If there are readability or other concerns let me know though and we can relook
